File: src/gapsense/curriculum/coverage.py
"""Manifest-driven curriculum coverage without filesystem inference."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from pathlib import Path

# FastAPI resolves this nested response type at runtime.
from gapsense.curriculum.inventory import SourceInventoryRecord  # noqa: TC001
from gapsense.curriculum.release import (
    CatalogCell,
    ReleaseManifest,
    ReleaseManifestError,
    ReleaseRecord,
    load_release_manifest,
)

RepositoryStatus = Literal["available", "partial", "missing", "invalid"]
AvailabilityStatus = Literal["present_unverified", "missing"]
ReviewStatus = Literal["not_verified", "human_reviewed"]
CoverageMatrixStatus = Literal[
    "missing",
    "located",
    "extracted",
    "structurally_validated",
    "human_reviewed",
]
EvidenceScope = Literal["level"]


@dataclass(frozen=True, slots=True)
class EducationLevel:
    """One official education level named by the country authority."""

    identifier: str
    name: str
    official_phase: str
    scope_note: str = "Current scope note unavailable until a release catalogue validates."
    review_status: ReviewStatus = "not_verified"


@dataclass(frozen=True, slots=True)
class CurriculumSubject:
    """One subject explicitly named by at least one release record."""

    identifier: str
    name: str
    phase: str
    availability: AvailabilityStatus
    review_status: ReviewStatus


@dataclass(frozen=True, slots=True)
class CoverageMatrixEntry:
    """One exact level/subject claim from the release manifest."""

    level_identifier: str
    level_name: str
    phase: str
    subject_identifier: str
    subject_name: str
    status: CoverageMatrixStatus
    evidence_scope: EvidenceScope
    source_url: str


@dataclass(frozen=True, slots=True)
class CountryDefinition:
    """Stable country and authority metadata controlled by the platform."""

    code: Literal["GH", "UG"]
    slug: Literal["ghana", "uganda"]
    name: Literal["Ghana", "Uganda"]
    authority: str
    authority_url: str
    levels: tuple[EducationLevel, ...]


@dataclass(frozen=True, slots=True)
class CountryCoverage:
    """Manifest evidence for one country."""

    code: Literal["GH", "UG"]
    name: Literal["Ghana", "Uganda"]
    authority: str
    authority_url: str
    availability: AvailabilityStatus
    review_status: ReviewStatus
    repository_file_count: int
    levels: tuple[EducationLevel, ...]
    subjects: tuple[CurriculumSubject, ...]
    coverage_matrix: tuple[CoverageMatrixEntry, ...]


@dataclass(frozen=True, slots=True)
class CoverageSnapshot:
    """The immutable release identity used by this application instance."""

    generated_at: str
    source_version: str | None
    review_status: ReviewStatus


@dataclass(frozen=True, slots=True)
class CatalogCoverage:
    """Representation and evidence counts for the pinned official catalogue."""

    as_of: str
    scope_status: Literal["official_authority_inventory"]
    represented_cells: int
    total_cells: int
    evidence_cells: int


@dataclass(frozen=True, slots=True)
class SourceInventoryCoverage:
    """Every sanitized official-source record pinned by the data release."""

    as_of: str
    total_records: int
    acquired_artifacts: int
    records: tuple[SourceInventoryRecord, ...]


@dataclass(frozen=True, slots=True)
class CoverageReport:
    """Deterministic two-country release coverage."""

    repository_status: RepositoryStatus
    complete: Literal[False]
    countries: tuple[CountryCoverage, ...]
    warnings: tuple[str, ...]
    snapshot: CoverageSnapshot
    catalog: CatalogCoverage | None
    source_inventory: SourceInventoryCoverage | None


COUNTRY_DEFINITIONS: tuple[CountryDefinition, ...] = (
    CountryDefinition(
        code="GH",
        slug="ghana",
        name="Ghana",
        authority="National Council for Curriculum and Assessment (NaCCA)",
        authority_url="https://nacca.gov.gh/curriculum/",
        levels=(
            EducationLevel("kindergarten", "Kindergarten", "Key Phase 1"),
            EducationLevel("lower_primary", "Lower Primary", "Key Phase 2 (Basic 1–3)"),
            EducationLevel("upper_primary", "Upper Primary", "Key Phase 3 (Basic 4–6)"),
            EducationLevel("junior_high", "JHS (Basic 7–9)", "Key Phase 4"),
            EducationLevel("senior_high", "SHS", "Key Phase 5"),
        ),
    ),
    CountryDefinition(
        code="UG",
        slug="uganda",
        name="Uganda",
        authority="National Curriculum Development Centre (NCDC)",
        authority_url="https://ncdc.go.ug/directorates/",
        levels=(
            EducationLevel("early_childhood", "Early Childhood", "ECCE"),
            EducationLevel("primary_1_3", "Primary One–Three", "Primary Phase 1"),
            EducationLevel("primary_4", "Primary Four", "Primary Phase 2 transition"),
            EducationLevel("primary_5_7", "Primary Five–Seven", "Primary Phase 3"),
            EducationLevel("lower_secondary", "O-Level (S1–S4)", "UCE cycle"),
            EducationLevel("upper_secondary", "A-Level (S5–S6)", "UACE cycle"),
        ),
    ),
)


def _new_snapshot(source_version: str | None) -> CoverageSnapshot:
    generated_at = datetime.now(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")
    return CoverageSnapshot(
        generated_at=generated_at,
        source_version=source_version,
        review_status="not_verified",
    )


def _subject_records(
    records: tuple[ReleaseRecord, ...],
    cells: tuple[CatalogCell, ...],
) -> tuple[CurriculumSubject, ...]:
    grouped: dict[tuple[str, str], list[CatalogCell]] = {}
    for cell in cells:
        grouped.setdefault((cell.phase, cell.subject), []).append(cell)
    record_groups: dict[tuple[str, str], list[ReleaseRecord]] = {}
    for record in records:
        record_groups.setdefault((record.phase, record.subject), []).append(record)
    subjects = []
    for (phase, identifier), catalog_matches in grouped.items():
        record_matches = record_groups.get((phase, identifier), [])
        subjects.append(
            CurriculumSubject(
                identifier=identifier,
                name=catalog_matches[0].subject_name,
                phase=phase,
                availability=(
                    "present_unverified"
                    if any(record.maturity != "missing" for record in record_matches)
                    else "missing"
                ),
                review_status=(
                    "human_reviewed"
                    if record_matches
                    and all(
                        record.review_status == "human_reviewed"
                        for record in record_matches
                        if record.maturity != "missing"
                    )
                    and any(record.maturity != "missing" for record in record_matches)
                    else "not_verified"
                ),
            )
        )
    return tuple(sorted(subjects, key=lambda subject: (subject.phase, subject.identifier)))


def _country_coverage(
    country: CountryDefinition,
    manifest: ReleaseManifest,
) -> CountryCoverage:
    records = tuple(record for record in manifest.records if record.country == country.code)
    cells = tuple(cell for cell in manifest.catalog_cells if cell.country == country.code)
    record_index = {(record.phase, record.level, record.subject): record for record in records}
    projection_paths = {
        path
        for record in records
        if record.projection is not None
        for path in (record.projection.nodes_path, record.projection.graph_path)
    }
    matrix = tuple(
        CoverageMatrixEntry(
            level_identifier=cell.level,
            level_name=cell.level_name,
            phase=cell.phase,
            subject_identifier=cell.subject,
            subject_name=cell.subject_name,
            status=(
                record_index[(cell.phase, cell.level, cell.subject)].maturity
                if (cell.phase, cell.level, cell.subject) in record_index
                else "missing"
            ),
            evidence_scope="level",
            source_url=cell.source_url,
        )
        for cell in sorted(
            cells,
            key=lambda item: (item.phase, item.level, item.subject),
        )
    )
    available_records = tuple(record for record in records if record.maturity != "missing")
    official_phases = {level.identifier: level.official_phase for level in country.levels}
    observed_levels: set[str] = set()
    catalog_levels = []
    for cell in cells:
        if cell.level in observed_levels:
            continue
        observed_levels.add(cell.level)
        catalog_levels.append(
            EducationLevel(
                identifier=cell.level,
                name=cell.level_name,
                official_phase=official_phases[cell.level],
                scope_note=cell.scope_note,
            )
        )
    return CountryCoverage(
        code=country.code,
        name=country.name,
        authority=country.authority,
        authority_url=country.authority_url,
        availability="present_unverified" if available_records else "missing",
        review_status=(
            "human_reviewed"
            if available_records
            and all(record.review_status == "human_reviewed" for record in available_records)
            else "not_verified"
        ),
        repository_file_count=len(projection_paths),
        levels=tuple(catalog_levels),
        subjects=_subject_records(records, cells),
        coverage_matrix=matrix,
    )


def _unavailable_report(
    status: Literal["missing", "invalid"],
    warning: str,
) -> CoverageReport:
    return CoverageReport(
        repository_status=status,
        complete=False,
        countries=tuple(
            CountryCoverage(
                code=country.code,
                name=country.name,
                authority=country.authority,
                authority_url=country.authority_url,
                availability="missing",
                review_status="not_verified",
                repository_file_count=0,
                levels=country.levels,
                subjects=(),
                coverage_matrix=(),
            )
            for country in COUNTRY_DEFINITIONS
        ),
        warnings=(warning,),
        snapshot=_new_snapshot(None),
        catalog=None,
        source_inventory=None,
    )


def canonical_repository_available(data_path: Path) -> bool:
    """Return whether one explicit release manifest and its pinned bytes validate."""
    try:
        load_release_manifest(data_path)
    except ReleaseManifestError:
        return False
    return True


def build_coverage_report(
    data_path: Path,
    *,
    manifest: ReleaseManifest | None = None,
) -> CoverageReport:
    """Build coverage exclusively from one byte-verified release manifest."""
    if manifest is None:
        release_path = data_path / "releases" / "curriculum-release.json"
        if release_path.is_symlink() or not release_path.is_file():
            return _unavailable_report("missing", "missing_release_manifest")
        try:
            manifest = load_release_manifest(data_path)
        except ReleaseManifestError:
            return _unavailable_report("invalid", "invalid_release_manifest")

    warning = {
        "candidate": "candidate_release",
        "empty": "empty_release",
        "released": None,
    }[manifest.release_status]
    return CoverageReport(
        repository_status="available",
        complete=False,
        countries=tuple(_country_coverage(country, manifest) for country in COUNTRY_DEFINITIONS),
        warnings=(warning,) if warning is not None else (),
        snapshot=_new_snapshot(manifest.release_id),
        catalog=CatalogCoverage(
            as_of=manifest.catalog_as_of,
            scope_status=manifest.catalog_scope_status,
            represented_cells=len(manifest.catalog_cells),
            total_cells=len(manifest.catalog_cells),
            evidence_cells=sum(record.maturity != "missing" for record in manifest.records),
        ),
        source_inventory=SourceInventoryCoverage(
            as_of=manifest.source_inventory.as_of,
            total_records=len(manifest.source_inventory.records),
            acquired_artifacts=sum(
                record.artifact_available for record in manifest.source_inventory.records
            ),
            records=manifest.source_inventory.records,
        ),
    )

File: frontend/src/services/coverage.ts
const coverageTimeoutMilliseconds = 5_000;

export type RepositoryStatus = "available" | "partial" | "missing" | "invalid";
export type AvailabilityStatus = "present_unverified" | "missing";
export type ReviewStatus = "not_verified" | "human_reviewed";

export type CurriculumLevel = {
  readonly identifier: string;
  readonly name: string;
  readonly official_phase: string;
  readonly scope_note: string;
  readonly review_status: ReviewStatus;
};

export type CurriculumSubject = {
  readonly identifier: string;
  readonly name: string;
  readonly phase: string;
  readonly availability: AvailabilityStatus;
  readonly review_status: ReviewStatus;
};

export type CoverageMatrixEntry = {
  readonly level_identifier: string;
  readonly level_name: string;
  readonly phase: string;
  readonly subject_identifier: string;
  readonly subject_name: string;
  readonly status:
    "missing" | "located" | "extracted" | "structurally_validated" | "human_reviewed";
  readonly evidence_scope: "level";
  readonly source_url: string;
};

export type CountryCoverage = {
  readonly code: "GH" | "UG";
  readonly name: "Ghana" | "Uganda";
  readonly authority: string;
  readonly authority_url: string;
  readonly availability: AvailabilityStatus;
  readonly review_status: ReviewStatus;
  readonly repository_file_count: number;
  readonly levels: readonly CurriculumLevel[];
  readonly subjects: readonly CurriculumSubject[];
  readonly coverage_matrix: readonly CoverageMatrixEntry[];
};

export type CoverageSnapshot = {
  readonly generated_at: string;
  readonly source_version: string | null;
  readonly review_status: ReviewStatus;
};

export type CatalogCoverage = {
  readonly as_of: string;
  readonly scope_status: "official_authority_inventory";
  readonly represented_cells: number;
  readonly total_cells: number;
  readonly evidence_cells: number;
};

export type SourceInventoryRecord = {
  readonly identifier: string;
  readonly country: "GH" | "UG";
  readonly phase: "primary" | "secondary";
  readonly level: string;
  readonly subject: string;
  readonly edition: string;
  readonly source_url: string;
  readonly retrieved_on: string;
  readonly license_status:
    | "all_rights_reserved_permission_required"
    | "official_index_only_document_not_licensed_for_redistribution"
    | "restricted_internal_research_permission_required_for_redistribution";
  readonly artifact_available: boolean;
  readonly artifact_pages: number | null;
  readonly extraction_status:
    "index_only" | "not_extracted" | "normalized_projection_unverified" | "text_present_unverified";
  readonly review_status:
    | "legacy_artifact_byte_verified_provenance_reacquisition_pending"
    | "official_artifact_byte_verified"
    | "official_artifact_byte_verified_document_phase_review_pending"
    | "official_document_identity_pending_binary_comparison"
    | "official_index_and_framework_verified"
    | "official_index_verified";
  readonly known_gap: string;
};

export type SourceInventoryCoverage = {
  readonly as_of: string;
  readonly total_records: number;
  readonly acquired_artifacts: number;
  readonly records: readonly SourceInventoryRecord[];
};

export type CurriculumCoverageReport = {
  readonly repository_status: RepositoryStatus;
  readonly complete: false;
  readonly countries: readonly CountryCoverage[];
  readonly warnings: readonly string[];
  readonly snapshot: CoverageSnapshot;
  readonly catalog: CatalogCoverage | null;
  readonly source_inventory: SourceInventoryCoverage | null;
};

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null;

const isReviewStatus = (value: unknown): value is ReviewStatus =>
  value === "not_verified" || value === "human_reviewed";

const isLevel = (value: unknown): value is CurriculumLevel =>
  isRecord(value) &&
  typeof value.identifier === "string" &&
  typeof value.name === "string" &&
  typeof value.official_phase === "string" &&
  typeof value.scope_note === "string" &&
  value.scope_note.trim() !== "" &&
  isReviewStatus(value.review_status);

const hasExpectedCountryIdentity = (value: Record<string, unknown>): boolean =>
  (value.code === "GH" && value.name === "Ghana") ||
  (value.code === "UG" && value.name === "Uganda");

const isLevelCatalog = (value: unknown): value is readonly CurriculumLevel[] =>
  Array.isArray(value) &&
  value.length > 0 &&
  value.every(isLevel) &&
  new Set(value.map((level) => level.identifier)).size === value.length;

const isSubject = (value: unknown): value is CurriculumSubject =>
  isRecord(value) &&
  typeof value.identifier === "string" &&
  typeof value.name === "string" &&
  typeof value.phase === "string" &&
  (value.availability === "present_unverified" || value.availability === "missing") &&
  isReviewStatus(value.review_status);

const isSubjectCatalog = (value: unknown): value is readonly CurriculumSubject[] =>
  Array.isArray(value) &&
  value.every(isSubject) &&
  new Set(value.map((subject) => `${subject.phase}:${subject.identifier}`)).size === value.length;

const isCoverageMatrixEntry = (value: unknown): value is CoverageMatrixEntry =>
  isRecord(value) &&
  typeof value.level_identifier === "string" &&
  typeof value.level_name === "string" &&
  typeof value.phase === "string" &&
  typeof value.subject_identifier === "string" &&
  typeof value.subject_name === "string" &&
  (
    ["missing", "located", "extracted", "structurally_validated", "human_reviewed"] as const
  ).includes(value.status as CoverageMatrixEntry["status"]) &&
  value.evidence_scope === "level" &&
  typeof value.source_url === "string";

const isCoverageMatrix = (value: unknown): value is readonly CoverageMatrixEntry[] =>
  Array.isArray(value) &&
  value.every(isCoverageMatrixEntry) &&
  new Set(
    value.map((entry) => `${entry.phase}:${entry.level_identifier}:${entry.subject_identifier}`),
  ).size === value.length;

const utcTimestamp = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;

const isCoverageSnapshot = (value: unknown): value is CoverageSnapshot =>
  isRecord(value) &&
  typeof value.generated_at === "string" &&
  utcTimestamp.test(value.generated_at) &&
  (value.source_version === null ||
    (typeof value.source_version === "string" && value.source_version.length > 0)) &&
  isReviewStatus(value.review_status);

const isCountry = (value: unknown): value is CountryCoverage => {
  if (
    !isRecord(value) ||
    !hasExpectedCountryIdentity(value) ||
    typeof value.authority !== "string" ||
    typeof value.authority_url !== "string" ||
    (value.availability !== "present_unverified" && value.availability !== "missing") ||
    !isReviewStatus(value.review_status) ||
    !Number.isSafeInteger(value.repository_file_count) ||
    Number(value.repository_file_count) < 0 ||
    !isLevelCatalog(value.levels) ||
    !isSubjectCatalog(value.subjects) ||
    !isCoverageMatrix(value.coverage_matrix)
  ) {
    return false;
  }

  const levels = new Map(value.levels.map((level) => [level.identifier, level]));
  const subjects = new Map(
    value.subjects.map((subject) => [`${subject.phase}:${subject.identifier}`, subject]),
  );
  for (const entry of value.coverage_matrix) {
    const level = levels.get(entry.level_identifier);
    const subject = subjects.get(`${entry.phase}:${entry.subject_identifier}`);
    const officialSourcePrefix =
      value.code === "GH" ? "https://nacca.gov.gh/" : "https://ncdc.go.ug/";
    if (
      level?.name !== entry.level_name ||
      subject?.name !== entry.subject_name ||
      !entry.source_url.startsWith(officialSourcePrefix)
    ) {
      return false;
    }
  }

  for (const subject of value.subjects) {
    const records = value.coverage_matrix.filter(
      (entry) => entry.phase === subject.phase && entry.subject_identifier === subject.identifier,
    );
    const expectedAvailability = records.some((entry) => entry.status !== "missing")
      ? "present_unverified"
      : "missing";
    if (records.length === 0 || subject.availability !== expectedAvailability) {
      return false;
    }
  }

  const expectedCountryAvailability = value.coverage_matrix.some(
    (entry) => entry.status !== "missing",
  )
    ? "present_unverified"
    : "missing";
  return value.availability === expectedCountryAvailability;
};

const isoDate = /^\d{4}-\d{2}-\d{2}$/;

const isCatalogCoverage = (value: unknown): value is CatalogCoverage =>
  isRecord(value) &&
  typeof value.as_of === "string" &&
  isoDate.test(value.as_of) &&
  value.scope_status === "official_authority_inventory" &&
  Number.isSafeInteger(value.represented_cells) &&
  Number(value.represented_cells) > 0 &&
  Number.isSafeInteger(value.total_cells) &&
  value.total_cells === value.represented_cells &&
  Number.isSafeInteger(value.evidence_cells) &&
  Number(value.evidence_cells) >= 0 &&
  Number(value.evidence_cells) <= Number(value.total_cells);

const isSourceInventoryRecord = (value: unknown): value is SourceInventoryRecord => {
  if (
    !isRecord(value) ||
    (value.country !== "GH" && value.country !== "UG") ||
    (value.phase !== "primary" && value.phase !== "secondary") ||
    typeof value.identifier !== "string" ||
    typeof value.level !== "string" ||
    typeof value.subject !== "string" ||
    typeof value.edition !== "string" ||
    typeof value.source_url !== "string" ||
    typeof value.retrieved_on !== "string" ||
    !isoDate.test(value.retrieved_on) ||
    !(
      [
        "all_rights_reserved_permission_required",
        "official_index_only_document_not_licensed_for_redistribution",
        "restricted_internal_research_permission_required_for_redistribution",
      ] as const
    ).includes(value.license_status as SourceInventoryRecord["license_status"]) ||
    typeof value.artifact_available !== "boolean" ||
    !(
      value.artifact_pages === null ||
      (Number.isSafeInteger(value.artifact_pages) && Number(value.artifact_pages) > 0)
    ) ||
    !(
      [
        "index_only",
        "not_extracted",
        "normalized_projection_unverified",
        "text_present_unverified",
      ] as const
    ).includes(value.extraction_status as SourceInventoryRecord["extraction_status"]) ||
    !(
      [
        "legacy_artifact_byte_verified_provenance_reacquisition_pending",
        "official_artifact_byte_verified",
        "official_artifact_byte_verified_document_phase_review_pending",
        "official_document_identity_pending_binary_comparison",
        "official_index_and_framework_verified",
        "official_index_verified",
      ] as const
    ).includes(value.review_status as SourceInventoryRecord["review_status"]) ||
    typeof value.known_gap !== "string"
  ) {
    return false;
  }
  const sourcePrefix = value.country === "GH" ? "https://nacca.gov.gh/" : "https://ncdc.go.ug/";
  return (
    value.source_url.startsWith(sourcePrefix) &&
    value.artifact_available === (value.artifact_pages !== null)
  );
};

const isSourceInventory = (value: unknown): value is SourceInventoryCoverage =>
  isRecord(value) &&
  typeof value.as_of === "string" &&
  isoDate.test(value.as_of) &&
  Number.isSafeInteger(value.total_records) &&
  Number(value.total_records) > 0 &&
  Number.isSafeInteger(value.acquired_artifacts) &&
  Number(value.acquired_artifacts) >= 0 &&
  Array.isArray(value.records) &&
  value.records.length === value.total_records &&
  value.records.every(isSourceInventoryRecord) &&
  value.records.every((record) => record.retrieved_on <= (value.as_of as string)) &&
  new Set(value.records.map((record) => record.identifier)).size === value.records.length &&
  value.records.filter((record) => record.artifact_available).length === value.acquired_artifacts;

const isCoverageReport = (value: unknown): value is CurriculumCoverageReport => {
  if (!isRecord(value) || value.complete !== false) {
    return false;
  }
  if (
    !(["available", "partial", "missing", "invalid"] as const).includes(
      value.repository_status as RepositoryStatus,
    )
  ) {
    return false;
  }
  const hasRelease =
    value.repository_status === "available" || value.repository_status === "partial";
  if (
    (hasRelease &&
      (!isCatalogCoverage(value.catalog) || !isSourceInventory(value.source_inventory))) ||
    (!hasRelease && (value.catalog !== null || value.source_inventory !== null))
  ) {
    return false;
  }
  if (
    !Array.isArray(value.warnings) ||
    !value.warnings.every((warning) => typeof warning === "string") ||
    !isCoverageSnapshot(value.snapshot)
  ) {
    return false;
  }
  if (
    (hasRelease && value.snapshot.source_version === null) ||
    ((value.repository_status === "missing" || value.repository_status === "invalid") &&
      value.snapshot.source_version !== null)
  ) {
    return false;
  }
  if (
    !Array.isArray(value.countries) ||
    value.countries.length !== 2 ||
    !value.countries.every(isCountry)
  ) {
    return false;
  }

  if (new Set(value.countries.map((country) => country.code)).size !== 2) {
    return false;
  }
  if (isCatalogCoverage(value.catalog)) {
    const entries = value.countries.flatMap((country) => country.coverage_matrix);
    return (
      entries.length === value.catalog.total_cells &&
      entries.filter((entry) => entry.status !== "missing").length === value.catalog.evidence_cells
    );
  }
  return value.countries.every(
    (country) => country.subjects.length === 0 && country.coverage_matrix.length === 0,
  );
};

export async function getCurriculumCoverage(
  fetcher: typeof fetch = fetch,
): Promise<CurriculumCoverageReport> {
  const response = await fetcher("/api/v1/curriculum/coverage", {
    headers: { Accept: "application/json" },
    signal: AbortSignal.timeout(coverageTimeoutMilliseconds),
  });

  if (!response.ok) {
    throw new Error(`coverage endpoint returned ${String(response.status)}`);
  }

  const payload: unknown = await response.json();
  if (!isCoverageReport(payload)) {
    throw new Error("coverage endpoint returned an invalid payload");
  }

  return payload;
}
